### DISCUSSION TITLE
Document SQL codegen, trade-offs

### DISCUSSION THREAD

**Discussion by jods4** [geldata/gel#3468] [General]
I just found out about EdgeDB and I'm intrigued!
The idea of modernizing the model and query language of RDBMS is excellent and challenges the benefits of NoSQL alternatives.

That said, data is hard and the _status quo_ exists because there are many trade-offs in implementation choices.

Before adopting EdgeDB, I would be genuinely interested in understanding exactly what choices have been made for me.
I suppose this could be a whitepaper that explains how the modelling and queries are translated into SQL.

Some examples:
- Decisions such as SERIALIZABLE only transactions #3466 are non-trivial;
- How do you fetch data when there are multiple related collections with cardinality > 1?
- How is inheritance modelled? TPT, TPC, TPH all have different pros and cons. I assumed you might use Postgre built-in inheritance (which has its own limitations -- to my understanding is TPC but I'm not a Postgre expert) but #3290 seems to indicate you're not.
- What are the semantics of accessing linked collections? The docs gives a sample translation of INNER JOIN, but the answer to #3460 says it's a LEFT JOIN (makes more sense). Can you do FULL OUTER? How to express INNER JOIN?
- Etc. there are many interesting topics like those ones.

**Comment by elprans** [comment]
Very good questions, thanks @jods4!

> I suppose this could be a whitepaper that explains how the modelling and queries are translated into SQL.

Yes, we don't intend to make anything a secret, so lack of deep technical documentation is simply due to lack of capacity.  @msullivan explained some of the internals in his [launch day talk](https://youtu.be/WRZ3o-NsU_4?t=8150).

> Decisions such as SERIALIZABLE only transactions

I replied in #3466

> How do you fetch data when there are multiple related collections with cardinality > 1?

It depends.  If you want to _return_ a related collection, it's roughly (`select array_agg(coll_query) from source`), in filters and traversals it's semi-joins and anti-joins (`select foo where bar in (select .. multi_link_rel)`).

> How is inheritance modelled?

We used to use native inheritance, but as you pointed out, it has a bunch of limitations (incompatibility with native partitioning is the worst), so now we simply maintain views over a union of descendant tables (TPC in the parlance), which results in the exact same plan as scanning an inherited table: a bunch of (possibly parallel) `Append` nodes.  Mutations are handled by the compiler directly, i.e. `update ParentType` is compiled into a bunch of `UPDATE` CTEs for the inheritance hierarchy.

> What are the semantics of accessing linked collections?

Linked collections are modeled as `(source_id, target_id [, <link properties>...])` just like you would expect them to.

> Can you do FULL OUTER? How to express INNER JOIN?

Like I mentioned in the sibling discussion, it's quite hard to reason about these without discussing concrete queries.  What would a `FULL OUTER JOIN` for a linked collection even mean?  EdgeQL isn't SQL and so certain patterns would not reflect 100%, but obviously there should be a way to express what you _want_ (or it's a bug).

Finally, some things to consider:

1. EdgeDB targets PostgreSQL _exclusively_, and so the query compiler is free to choose the most optimal formulation for a query pattern regardless of how arcane or Postgres-specific it is in SQL terms.

2. There is already ~40K of Python code in the query compiler (that's not including the schema logic), and a lot of it due to a large number of specializations required to generate optimal SQL for a variety of patterns.

**Reply by amaster507** [reply]
I want to try to tackle the `OUTER JOIN` question too as I had to cross that bridge when _limiting_ my FE to only be able to use GraphQL and not creating a BE (which I am now looking to build on top of edgeDB). The key phrase to answer the join understanding is
> Right joins are just left joins in reverse.
👀 for a reply on the other topic soon with some concrete side-by-side comparisons of join SQL syntax to edgeQL syntax. I learn best by trying to answer the tough questions.

**Reply by jods4** [reply]
Thanks, that's instructive.

It's too late now but I sure will look at the talk.

The array_agg strategy is really interesting -- and unusual, as it's not a SQL commonplace!

I look forward to learning more if you get to write some docs about the EdgeDB translations.

> EdgeDB targets PostgreSQL exclusively, and so the query compiler is free to choose the most optimal formulation for a query pattern regardless of how arcane or Postgres-specific it is in SQL terms.

I can see why!

There would be a lot of value in being able to plug EdgeDB on top of any legacy DB you have, but it's a _massive_ endeavor.
- You'd need a lot more flexibility in schema modelling, e.g. supporting identity for ids rather than guids;
- You'd need to work out provider-specific translations of everything, e.g. array_agg is not an option with Oracle.

Sounds like a whole lot of work.
Postgres is a great choice, as it has awesome feature support and is free/open-source.
